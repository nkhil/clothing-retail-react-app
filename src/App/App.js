import React, { Component, Fragment } from "react"
import products from "./helpers/products"
import voucherCodes from "./helpers/voucherCodes"
import ProductList from "./components/ProductList"
import ShoppingCart from "./components/Cart"

class App extends Component {
  state = {
    products: {},
    shoppingCart: {},
    voucherCodes: [],
    activeVoucherCode: null,
  }

  componentWillMount = () => {
    this.loadProductsIntoState(products)
    this.loadVoucherCodesIntoState(voucherCodes)
  }

  loadProductsIntoState = products => {
    this.setState({ products })
  }

  loadVoucherCodesIntoState = voucherCodes => {
    this.setState({ voucherCodes })
  }

  addToCart = key => {
    const shoppingCart = { ...this.state.shoppingCart }
    shoppingCart[key] = shoppingCart[key] + 1 || 1
    this.setState({ shoppingCart })
  }

  removeFromCart = key => {
    const shoppingCart = { ...this.state.shoppingCart }
    shoppingCart[key] = shoppingCart[key] - 1
    if (shoppingCart[key] <= 0) {
      delete shoppingCart[key]
    }
    this.setState({ shoppingCart })
  }

  deductProductFromInventory = key => {
    let products = { ...this.state.products }
    products[key].productQuantity -= 1
    this.setState({ products })
  }

  calculateTotal = () => {
    const cartIds = Object.keys(this.state.shoppingCart)
    const cartTotal = cartIds.reduce((prevTotal, key) => {
      const productPrice = this.state.products[key].productPrice
      const count = this.state.shoppingCart[key]
      return prevTotal + productPrice * count
    }, 0)
    return cartTotal
  }

  calculateDiscountedTotal = () => {
    if (this.state.activeVoucherCode) {
      const discountAmount = this._getVoucherObject(
        this.state.activeVoucherCode
      ).discountAmount
      const discountedTotal = this.calculateTotal() - discountAmount
      return discountedTotal
    }
    return this.calculateTotal()
  }

  setActiveVoucherCode = voucherCode => {
    this.setState({ activeVoucherCode: voucherCode })
  }

  voucherCodeIsValid = voucherCode => {
    const voucherObject = this._getVoucherObject(voucherCode)
    if (voucherObject && voucherCode === "FIFTEENOFF") {
      return this._checkFifteenOffCriteria(voucherObject.minimumSpend)
    }
    if (voucherObject) {
      return this._voucherMeetsMinimumSpend(voucherObject.minimumSpend)
    }
    return false
  }

  _checkFifteenOffCriteria = minimumSpend => {
    return (
      this._voucherMeetsMinimumSpend(minimumSpend) &&
      this._shoppingCartContainsFootwear()
    )
  }

  _shoppingCartContainsFootwear = () => {
    const shoppingCartIds = Object.keys(this.state.shoppingCart)
    const footwearItemsInCart = shoppingCartIds.filter(key => {
      return this.state.products[key].productCategory.includes("Footwear")
    })
    return footwearItemsInCart.length > 0
  }

  _voucherMeetsMinimumSpend = minimumSpend =>
    this.calculateDiscountedTotal() >= minimumSpend

  _getVoucherObject = voucherCode => {
    const voucherObject = this.state.voucherCodes.filter(
      voucher => voucher.code === voucherCode
    )
    return voucherObject[0]
  }

  render = () => {
    return (
      <Fragment>
        <ProductList
          products={this.state.products}
          addToCart={this.addToCart}
          deductProductFromInventory={this.deductProductFromInventory}
        />
        <ShoppingCart
          products={this.state.products}
          shoppingCart={this.state.shoppingCart}
          removeFromCart={this.removeFromCart}
          calculateDiscountedTotal={this.calculateDiscountedTotal}
          voucherCodeIsValid={this.voucherCodeIsValid}
          setActiveVoucherCode={this.setActiveVoucherCode}
        />
      </Fragment>
    )
  }
}

export default App
