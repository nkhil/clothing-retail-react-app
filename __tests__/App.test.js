import "@babel/polyfill"
import React from "react"
import App from "../src/App/App"
import ReactDOM from "react-dom"
import renderer from "react-test-renderer"

require("react-router-dom")
jest.mock("react-router-dom")

test("App renders without crashing", () => {
  const div = document.createElement("div")
  ReactDOM.render(<App />, div)
})

test("App renders correctly", () => {
  const app = renderer.create(<App />).toJSON()
  expect(app).toMatchSnapshot()
})

test("addToCart()", () => {
  const shoppingBasket = { product1: 1 }
  const component = renderer.create(<App />)
  const componentInstance = component.getInstance()
  componentInstance.addToCart("product1")
  expect(componentInstance.state.shoppingCart).toEqual(shoppingBasket)
})

test("deductProductFromInventory()", () => {
  const component = renderer.create(<App />)
  const componentInstance = component.getInstance()
  componentInstance.addToCart("product1")
  componentInstance.deductProductFromInventory("product1")
  expect(componentInstance.state.products["product1"].productQuantity).toEqual(
    4
  )
})

test("removeFromCart()", () => {
  const component = renderer.create(<App />)
  const componentInstance = component.getInstance()
  componentInstance.addToCart("product1")
  componentInstance.removeFromCart("product1")
  expect(componentInstance.state.shoppingCart).toEqual({})
})

test("calculateTotal()", () => {
  const component = renderer.create(<App />)
  const componentInstance = component.getInstance()
  componentInstance.addToCart("product2")
  expect(componentInstance.calculateTotal()).toEqual(4200)
})

test("calculateDiscountedTotal() in the presence of discount code", () => {
  const component = renderer.create(<App />)
  const componentInstance = component.getInstance()
  const mockVoucherCode = { code: "FIVEOFF", discountAmount: 500 }
  componentInstance.state.voucherCodes = [mockVoucherCode]
  componentInstance.setActiveVoucherCode("FIVEOFF")
  componentInstance.addToCart("product1")
  expect(componentInstance.calculateDiscountedTotal()).toEqual(9400)
})

test("calculateDiscountedTotal() in the absence of discount code", () => {
  const component = renderer.create(<App />)
  const componentInstance = component.getInstance()
  componentInstance.addToCart("product1")
  expect(componentInstance.calculateDiscountedTotal()).toEqual(9900)
})

test("voucherCodeIsValid() with valid voucher code", () => {
  const mockDiscountCode = {
    code: "FIVEOFF",
    discountAmount: 500,
    minimumSpend: 0,
  }
  const component = renderer.create(<App />)
  const componentInstance = component.getInstance()
  componentInstance.state.voucherCodes = [mockDiscountCode]
  expect(componentInstance.voucherCodeIsValid("FIVEOFF")).toEqual(true)
})

test("voucherCodeIsValid() with invalid voucher code", () => {
  const mockDiscountCode = {
    code: "TEST",
    discountAmount: 500,
    minimumSpend: 0,
  }
  const component = renderer.create(<App />)
  const componentInstance = component.getInstance()
  componentInstance.state.voucherCodes = [mockDiscountCode]
  expect(componentInstance.voucherCodeIsValid("TES")).toEqual(false)
})

test("setActiveVoucherCode()", () => {
  const component = renderer.create(<App />)
  const componentInstance = component.getInstance()
  const mockDiscountCode = { code: "TEST", discountAmount: 500 }
  componentInstance.state.voucherCodes = [mockDiscountCode]
  componentInstance.setActiveVoucherCode("TEST")
  expect(componentInstance.state.activeVoucherCode).toEqual("TEST")
})
