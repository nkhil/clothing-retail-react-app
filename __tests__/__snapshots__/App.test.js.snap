// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders correctly 1`] = `
Array [
  <ul>
    <li
      id="product1"
    >
      <h3>
        Almond Toe Court Shoes, Patent Black
      </h3>
      <h4>
        £99.00
      </h4>
      <p>
        Women's Footwear
      </p>
      <button
        className="product1-button"
        onClick={[Function]}
      >
        Add to cart
      </button>
    </li>
    <li
      id="product2"
    >
      <h3>
        Suede Shoes, Blue
      </h3>
      <h4>
        £42.00
      </h4>
      <p>
        Women's Footwear
      </p>
      <button
        className="product2-button"
        onClick={[Function]}
      >
        Add to cart
      </button>
    </li>
    <li
      id="product3"
    >
      <h3>
        Leather Driver Saddle Loafers, Tan
      </h3>
      <h4>
        £34.00
      </h4>
      <p>
        Men's Footwear
      </p>
      <button
        className="product3-button"
        onClick={[Function]}
      >
        Add to cart
      </button>
    </li>
    <li
      id="product4"
    >
      <h3>
        Flip Flops, Red
      </h3>
      <h4>
        £19.00
      </h4>
      <p>
        Men's Footwear
      </p>
      <button
        className="product4-button"
        onClick={[Function]}
      >
        Add to cart
      </button>
    </li>
    <li
      id="product5"
    >
      <h3>
        Flip Flops, Blue
      </h3>
      <h4>
        £19.00
      </h4>
      <p>
        Men's Footwear
      </p>
      <button
        className="product5-button"
        onClick={[Function]}
      >
        Add to cart
      </button>
    </li>
    <li
      id="product6"
    >
      <h3>
        Gold Button Cardigan, Black
      </h3>
      <h4>
        £167.00
      </h4>
      <p>
        Women’s Casualwear
      </p>
      <button
        className="product6-button"
        onClick={[Function]}
      >
        Add to cart
      </button>
    </li>
    <li
      id="product7"
    >
      <h3>
        Cotton Shorts, Medium Red
      </h3>
      <h4>
        £30.00
      </h4>
      <p>
        Women’s Casualwear
      </p>
      <button
        className="product7-button"
        onClick={[Function]}
      >
        Add to cart
      </button>
    </li>
    <li
      id="product8"
    >
      <h3>
        Fine Stripe Short Sleeve Shirt, Grey
      </h3>
      <h4>
        £49.99
      </h4>
      <p>
        Men’s Casualwear
      </p>
      <button
        className="product8-button"
        onClick={[Function]}
      >
        Add to cart
      </button>
    </li>
    <li
      id="product9"
    >
      <h3>
        Fine Stripe Short Sleeve Shirt, Green
      </h3>
      <h4>
        £39.99
      </h4>
      <p>
        Men’s Casualwear
      </p>
      <button
        className="product9-button"
        onClick={[Function]}
      >
        Add to cart
      </button>
    </li>
    <li
      id="product10"
    >
      <h3>
        Sharkskin Waistcoat, Charcoal
      </h3>
      <h4>
        £75.00
      </h4>
      <p>
        Men’s Formalwear
      </p>
      <button
        className="product10-button"
        onClick={[Function]}
      >
        Add to cart
      </button>
    </li>
    <li
      id="product11"
    >
      <h3>
        Bird Print Dress, Black
      </h3>
      <h4>
        £270.00
      </h4>
      <p>
        Women’s Formalwear
      </p>
      <button
        className="product11-button"
        onClick={[Function]}
      >
        Add to cart
      </button>
    </li>
    <li
      id="product12"
    >
      <h3>
        Mid Twist Cut­Out Dress, Pink
      </h3>
      <h4>
        £540.00
      </h4>
      <p>
        Women’s Formalwear
      </p>
      <button
        className="product12-button"
        onClick={[Function]}
      >
        Add to cart
      </button>
    </li>
  </ul>,
  <div
    id="shoppingCart"
  >
    <h2>
      Basket:
    </h2>
    <ul />
    <h3
      id="total"
    >
      Total: 
      £0.00
    </h3>
  </div>,
]
`;
